igvm/vm.py

```diff
@@ -534,46 +534,15 @@ def get_best_hypervisor(self, hv_states=['online']):
             if self.dataset_obj['xen_host'] == hypervisor['hostname']:
                 continue
 
-<<<<<<< HEAD
-        for hv in sorted(ranking, key=ranking.get, reverse=True):
-            if (
-                not self.dataset_obj['xen_host'] or
-                self.dataset_obj['xen_host'] != hv
-            ):
-                # Free disk size and memory are done in fast mode which might
-                # return wrong or orphaned values so to be sure we really have
-                # enough we check them here again.
-                bhv = BHypervisor(hv)
-                disk_free = bhv.get_disk_free() / 1024.0
-                if float(disk_free) > float(self.dataset_obj['disk_size_gib']):
-                    memory_free = bhv.get_memory_free()
-                    if memory_free > float(self.dataset_obj['memory']):
-                        return hv
-||||||| parent of 402bef9... balance: Remove ServeradminCache
-        for hv in sorted(ranking, key=ranking.get):
-            if (
-                not self.dataset_obj['xen_host'] or
-                self.dataset_obj['xen_host'] != hv
-            ):
-                # Free disk size and memory are done in fast mode which might
-                # return wrong or orphaned values so to be sure we really have
-                # enough we check them here again.
-                bhv = BHypervisor(hv)
-                disk_free = bhv.get_disk_free() / 1024.0
-                if float(disk_free) > float(self.dataset_obj['disk_size_gib']):
-                    memory_free = bhv.get_memory_free()
-                    if memory_free > float(self.dataset_obj['memory']):
-                        return hv
-=======
-            # Free disk size and memory are done in fast mode which might
-            # return wrong or orphaned values so to be sure we really have
-            # enough we check them here again.
-            disk_free = hypervisor.get_disk_free() / 1024.0
-            if float(disk_free) > float(self.dataset_obj['disk_size_gib']):
-                memory_free = hypervisor.get_memory_free()
-                if memory_free > float(self.dataset_obj['memory']):
-                    return hypervisor['hostname']
->>>>>>> 402bef9... balance: Remove ServeradminCache
+            # Free disk size and memory are done in fast mode on ranking.
+            # They might be wrong, so to be sure we really have enough,
+            # we check them here again.
+            if hypervisor.get_disk_free() <= self.dataset_obj['disk_size_gib']:
+                continue
+            if hypervisor.get_memory_free() <= self.dataset_obj['memory']:
+                continue
```

Comment:
get_disk_free() and get_memory_free() can now return None if the libvirt connection failed, which will raise a TypeError when compared with ints via the `<=` operator in python3.

Reply:
The methods of the `Hypervisor` class don't really return `None`.

Reply:
Well the try, finally block suggests that it can if ighv.free_vm_memory() raises an Exception.